docs/050-breaking-changes.rst

```diff
@@ -0,0 +1,206 @@
+  a ``bytes4`` (4 bytes) to a ``uint64`` (8 bytes) by first converting the
+  ``bytes4`` variable to ``bytes8`` and then to ``uint64``.
+
+* Using ``msg.value`` in ``non-payable`` functions (or introducing it via a
+  modifier) is disallowed as a security feature. Turn the function into
+  ``payable`` or create a new internal function for the program logic that
+  uses ``msg.value``.
+
+Example
+=======
+
+The following example shows a contract and its updated version for Solidity
+v0.5.0.
+Old version:
+
+::
+
+   pragma solidity ^0.4.24;
+
+   contract OtherContract {
+      uint x;
+      function f(uint y) external {
+         x = y;
+      }
+      function() payable external {}
+   }
+
+   contract Old {
+      OtherContract other;
+      uint myNumber;
+
+      // Function visibility not provided, not an error
+      // Function mutability not provided, not an error
+      function f(uint x) {
+         y = -3 >> 1; // y = -1 (wrong, should be -2)
+
+         do {
+            x += 1;
+            if (x > 10) continue;
+         } while (x < 11); // causes infinite loop
+
+         bool success = other.call("f"); // Call returns only a Bool
+         if (!success)
+            revert();
+         else {
+            int y; // Local variables could be declared after their use
+         }
+      }
+
+      function g(uint[] arr, bytes8 x, OtherContract otherContract) public { // No need to explicit parameter location data
+         otherContract.transfer(1 ether);
+
+         // Since uint32 (4 bytes) is smaller than bytes8 (8 bytes),
+         // the first 4 bytes of x will be lost. This is dangerous
+         // since bytesX are right padded.
+         uint32 y = uint32(x);
+         myNumber += y + msg.value;
+      }
+   }
+
+New version:
+
+::
+
+   pragma solidity >0.4.24;
+
+   contract OtherContract {
+      uint x;
+      function f(uint y) external {
+         x = y;
+      }
+      function() payable external {}
+   }
+
+   contract New {
+      OtherContract other;
+      uint myNumber;
+
+      // Function visibility not provided, not an error
+      // Function mutability not provided, not an error
+      function f(uint x) public returns (bytes memory) {
+         int y = -3 >> 1; // y = -2 (correct)
+
+         do {
+            x += 1;
+            if (x > 10) continue; // jumps to the condition below
+         } while (x < 11);
+
+         // Call returns (bool, bytes)
+         // Data location must be specified
+         (bool success, bytes memory data) = address(other).call("f");
+         if (!success)
+            revert();
+         return data;
+      }
+
+      using address_make_payable for address;
+      function g(uint[] memory arr, bytes8 x, OtherContract otherContract) public payable { // No need to explicit parameter location data
+         // 'contract.transfer' is not provided
+         // 'address(contract).transfer' is not provided since 'address(contract)' is not 'address payable'.
```

Comment:
Well, in this case ``address(otherContract)`` is in fact ``address payable``, since ``OtherContract`` *does* have a payable fallback function.

Reply:
Also: ``otherContract`` in code <=> ``contract`` in comment

Reply:
But does it already infer that? I'll change it to use a parameter `address` then

Reply:
Yes - ``address(contract)`` is of type ``address payable``, if ``contract`` has a payable fallback, and of type ``address``, if not.
So the only reason why one would use the ``address_make_payable`` library is, if someone gives you an ``address`` instead of an ``address payable``, although you know that it is in fact payable, but you can't change what you are given (e.g. because it's a return value from third party code you can't update). Maybe we should mention that here - we shouldn't promote using the workaround if not absolutely necessary...

Reply:
(Actually I'm not sure when one would ever have third party *source* code, which one can't update :-D)